vpr/src/base/partition_regions.h

```diff
@@ -0,0 +1,39 @@
+#ifndef PARTITION_REGIONS_H
+#define PARTITION_REGIONS_H
+
+#include "region.h"
+
+/**
+ * @file
+ * @brief This file defines the PartitionRegions class. The PartitionRegions class is used to store the union
+ * of regions that a partition can be placed in.
+ */
+
+class PartitionRegions {
+  public:
+    /**
+     * Returns the intersection of two PartitionRegions vectors that are passed to it.
+     */
+    PartitionRegions get_intersection(PartitionRegions region);
+
+    /**
+     * A test for intersection of the PartitionRegions of two blocks
+     */
+    bool intersects(PartitionRegions region);
+
+    /**
+     * Takes Atom Id returns intersection of atom's PartitionRegions with the PartitionRegions of the object
+     */
+    PartitionRegions get_intersect(const AtomBlockId atom_id);
+
+    //method to add to the partition_regions vector
+    void add_to_part_regions(Region region);
+
+    //method to get the partition_regions vector
+    std::vector<Region> get_partition_regions();
+
+  private:
+    std::vector<Region> partition_regions; //vector of regions that a partition can be placed in
```

Comment:
I'm not sure a vector of Region is the right data structure here.  I feel like you want a spatial tree, e.g. :
 - https://en.wikipedia.org/wiki/Binary_space_partitioning
 - https://en.wikipedia.org/wiki/K-d_tree

You probably want to prototype a couple different data structures balancing simplicity of implementation, data size overhead, etc.

At a minimum, I believe that PartitionRegion needs a `t_bb` that encompasses the vector of regions to provide a quick no if the query it outside any of the supplied regions.

Reply:
Common case is that the regions are simple (0 or 1 rectangle will be the dominant cases).  T-shaped and L-shaped will be possible but relatively rare.  So I think a k-d tree etc. is overkill and will be slower and more complex for the common case.

Reply:
Given your statement, should we consider a dynamic dispatch approach here?  Where the simple 0/1 case is implemented without the vector at all, and the more complicated case gets a more complicated object?

Reply:
We could.  I'd leave it simple for now though, and based on future need we can update if we hit cpu bottlenecks, memory footprint etc.  The class interface should hide all that from the callers, so re-coding would be limited.